File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/binary/PathProbeTest.kt
package com.tyhp.lang.binary

import java.nio.file.Files
import kotlin.io.path.writeText
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class PathProbeTest {
    private val unix = HostPlatform(OsId.OSX, ArchId.ARM64)
    private val win = HostPlatform(OsId.WIN, ArchId.X64)

    @Test
    fun `PATH probe finds tyhp in the first matching directory`() {
        val root = Files.createTempDirectory("tyhp-path-")
        try {
            val hitDir = root.resolve("bin")
            val missDir = root.resolve("empty")
            Files.createDirectories(hitDir)
            Files.createDirectories(missDir)
            val binary = hitDir.resolve("tyhp")
            binary.writeText("#!/bin/sh\n")
            binary.toFile().setExecutable(true)
            val found = probeTyhpOnPath(
                PathProbeOptions(
                    pathEnv = listOf(missDir, hitDir).joinToString(System.getProperty("path.separator")),
                    pathDelimiter = System.getProperty("path.separator"),
                    platform = unix,
                ),
            )
            assertEquals(binary.toAbsolutePath().normalize().toString(), found)
        } finally {
            root.toFile().deleteRecursively()
        }
    }

    @Test
    fun `PATH probe returns null when tyhp is absent`() {
        val found = probeTyhpOnPath(
            PathProbeOptions(
                pathEnv = "/no/such/tyhp-bin",
                pathDelimiter = ":",
                platform = unix,
            ),
        )
        assertNull(found)
    }

    @Test
    fun `Windows PATH probe prefers tyhp exe`() {
        val names = mutableListOf<String>()
        val found = probeTyhpOnPath(
            PathProbeOptions(
                pathEnv = "C:\\Tools",
                pathDelimiter = ";",
                platform = win,
                fs = PathProbeFs { filePath ->
                    names.add(filePath.toString())
                    filePath.fileName.toString().equals("tyhp.exe", ignoreCase = true)
                },
            ),
        )
        assertNotNull(found)
        assertTrue(found.endsWith("tyhp.exe"))
        assertTrue(names.any { it.endsWith("tyhp.exe") })
    }

    @Test
    fun `setting precedence explicit path is validated and command names resolve via PATH`() {
        val root = Files.createTempDirectory("tyhp-validate-")
        try {
            val binary = root.resolve("tyhp")
            binary.writeText("#!/bin/sh\n")
            binary.toFile().setExecutable(true)
            val ok = validateTyhpPath(binary.toString(), PathProbeOptions(platform = unix))
            assertTrue(ok.ok)
            assertEquals(binary.toRealPath().toString(), ok.absolutePath)

            val missing = validateTyhpPath(root.resolve("missing").toString(), PathProbeOptions(platform = unix))
            assertEquals(false, missing.ok)
            assertTrue(missing.message.orEmpty().contains("missing or is not a file"))

            val viaPath = lookUpCommandOnPath(
                "tyhp",
                PathProbeOptions(
                    pathEnv = root.toString(),
                    pathDelimiter = System.getProperty("path.separator"),
                    platform = unix,
                ),
            )
            assertEquals(binary.toAbsolutePath().normalize().toString(), viaPath)

            val cmd = validateTyhpPath(
                "tyhp",
                PathProbeOptions(
                    pathEnv = root.toString(),
                    pathDelimiter = System.getProperty("path.separator"),
                    platform = unix,
                ),
            )
            assertTrue(cmd.ok)
            assertEquals(binary.toRealPath().toString(), cmd.absolutePath)
        } finally {
            root.toFile().deleteRecursively()
        }
    }
}

File: tyhp-lang/vscode/src/binary/PathProbe.test.ts
import assert from "node:assert/strict";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { test } from "node:test";
import { lookUpCommandOnPath, probeTyhpOnPath, validateTyhpPath } from "./PathProbe";
import { HostPlatform } from "./platform";

const unix: HostPlatform = { os: "osx", arch: "arm64", nodePlatform: "darwin" };
const win: HostPlatform = { os: "win", arch: "x64", nodePlatform: "win32" };

test("PATH probe finds tyhp in the first matching directory", () => {
    const root = fs.mkdtempSync(path.join(os.tmpdir(), "tyhp-path-"));
    const hitDir = path.join(root, "bin");
    const missDir = path.join(root, "empty");
    fs.mkdirSync(hitDir);
    fs.mkdirSync(missDir);
    const binary = path.join(hitDir, "tyhp");
    fs.writeFileSync(binary, "#!/bin/sh\n");
    fs.chmodSync(binary, 0o755);
    try {
        const found = probeTyhpOnPath({
            pathEnv: [missDir, hitDir].join(path.delimiter),
            pathDelimiter: path.delimiter,
            platform: unix,
        });
        assert.equal(found, binary);
    } finally {
        fs.rmSync(root, { recursive: true, force: true });
    }
});

test("PATH probe returns undefined when tyhp is absent", () => {
    const found = probeTyhpOnPath({
        pathEnv: "/no/such/tyhp-bin",
        pathDelimiter: ":",
        platform: unix,
    });
    assert.equal(found, undefined);
});

test("Windows PATH probe prefers tyhp.exe", () => {
    const names: string[] = [];
    const found = probeTyhpOnPath({
        pathEnv: "C:\\Tools",
        pathDelimiter: ";",
        platform: win,
        fs: {
            existsSync(filePath) {
                names.push(filePath);
                return filePath.replace(/\//g, "\\").endsWith("\\tyhp.exe") || filePath.endsWith("/tyhp.exe");
            },
            statSync() {
                return { isFile: () => true };
            },
        },
    });
    assert.ok(found?.endsWith("tyhp.exe"));
    assert.ok(names.some((n) => n.endsWith("tyhp.exe")));
});

test("setting precedence: explicit path is validated; command names resolve via PATH", () => {
    const root = fs.mkdtempSync(path.join(os.tmpdir(), "tyhp-validate-"));
    const binary = path.join(root, "tyhp");
    fs.writeFileSync(binary, "#!/bin/sh\n");
    fs.chmodSync(binary, 0o755);
    try {
        const ok = validateTyhpPath(binary, { platform: unix });
        assert.equal(ok.ok, true);
        assert.equal(ok.absolutePath, fs.realpathSync(binary));

        const missing = validateTyhpPath(path.join(root, "missing"), { platform: unix });
        assert.equal(missing.ok, false);
        assert.match(missing.message ?? "", /missing or is not a file/);

        const viaPath = lookUpCommandOnPath("tyhp", {
            pathEnv: root,
            pathDelimiter: path.delimiter,
            platform: unix,
        });
        assert.equal(viaPath, binary);

        const cmd = validateTyhpPath("tyhp", {
            pathEnv: root,
            pathDelimiter: path.delimiter,
            platform: unix,
        });
        assert.equal(cmd.ok, true);
        assert.equal(cmd.absolutePath, fs.realpathSync(binary));
    } finally {
        fs.rmSync(root, { recursive: true, force: true });
    }
});
